### PR TITLE
Template system

### PR DESCRIPTION
**Pull request by davepinto**
Implemented ability to modify colors in the template file, using functions in util.py.
Currently, there is only support for the lighten_color, and darken_color function. To use them, use a substitution marker like this in your template:
`{background.darken_color(20)}`
Unfortunately, to keep backwards compatibility, you can not use float values as a function argument. Instead, you pass a number between 0 and 100 (or -0 and -100), that represents a percentage.

### CODE REVIEW

File: pywal/export.py [hunk 1]
```diff
@@ -12,8 +13,31 @@ def template(colors, input_file, output_file=None):
     """Read template file, substitute markers and
        save the file elsewhere."""
     template_data = util.read_file_raw(input_file)
-    template_data = "".join(template_data).format(**colors)
 
+    # Match all substitution markers
+    regex = r"(?<!{){[^{|}]+}"
```
Comment (dylanaraps) [inline pywal/export.py]: This can be moved to the next line (so it only takes up one line). `regex` is only used once.

File: pywal/util.py [hunk 2]
```diff
@@ -123,12 +123,16 @@ def rgb_to_hex(color):
 
 def darken_color(color, amount):
     """Darken a hex color."""
+    if not isinstance(amount, float):
```
Comment (dylanaraps) [inline pywal/util.py]: I can update the usage of these throughout the program to take ints if it means removing these lines from these function.

File: pywal/export.py (hunk 1 above)
Comment (davepinto) [reply pywal/export.py]: Ok, I've fixed this, and will upload soon.

File: pywal/util.py (hunk 2 above)
Comment (davepinto) [reply pywal/util.py]: That would be awesome, thanks. I'll remove it now. For the sake of backwards compatability, I haven't been able to make it accept float arguments yet. Everytime it hits a dot, in the string format, it looks for an attribute named like the text after the dot, so I end up with an error like: `colors has no attribute '5)'`. Ideally I'd like to make it accept floats too. Do you have any idea how to do that? Perferably avoiding having to do something like `{background.darken_color(0\.5)}`.

Also, I was thinking that maybe the color editing functions should be loaded from the scripts directory, that way, users can implement their own color modifying functions. Could open up some neat configs, like an organised rainbow theme down the line.

**Comment by dustinlacewell** [comment]
Please consider using a tried and true template system like Jinja2 which is extremely capable and would serve us well.

**Comment by dylanaraps** [comment]
Something similar has been merged.

**Pull request closed without merging**